File: Backend/Backend/Services/QuizService.cs
using Backend.Data;
using Backend.DTOs;
using Backend.Models;
using Backend.Services.Interfaces;
using Microsoft.EntityFrameworkCore;

namespace Backend.Services
{
    public class QuizService : IQuizService
    {
        private readonly ApplicationDbContext context;

        public QuizService(ApplicationDbContext context)
        {
            this.context = context;
        }

        public async Task<QuizResponseDTO> CreateQuizAsync(QuizCreateDTO dto)
        {
            var quiz = new Quiz
            {
                Title = dto.Title,
                Description = dto.Description,
                Category = dto.Category,
                Difficulty = dto.Difficulty,
                LimitTime = dto.LimitTime,
                Questions = dto.Questions.Select(q => new Question
                {
                    Text = q.Text,
                    QuestionType = q.QuestionType,
                    Points = q.Points,
                    ExpectedAnswer = q.ExpectedAnswer,
                    IsMultipleChoice = q.QuestionType == "MultipleChoice",
                    Options = q.Options.Select(o => new Option
                    {
                        Text = o.Text,
                        IsCorrect = o.IsCorrect
                    }).ToList()
                }).ToList()
            };

            context.Quizzes.Add(quiz);
            await context.SaveChangesAsync();

            return await GetQuizByIdAsync(quiz.Id) ?? throw new Exception("Failed to create quiz");
        }

        public async Task<QuizResponseDTO?> GetQuizByIdAsync(int id)
        {
            var quiz = await context.Quizzes
                .Include(q => q.Questions)
                    .ThenInclude(q => q.Options)
                .FirstOrDefaultAsync(q => q.Id == id);

            if (quiz == null)
                return null;

            return new QuizResponseDTO
            {
                Id = quiz.Id,
                Title = quiz.Title,
                Description = quiz.Description,
                Category = quiz.Category,
                Difficulty = quiz.Difficulty,
                LimitTime = quiz.LimitTime,
                Questions = quiz.Questions.Select(q => new QuestionResponseDTO
                {
                    Id = q.Id,
                    Text = q.Text,
                    QuestionType = q.QuestionType,
                    Points = q.Points,
                    ExpectedAnswer = q.ExpectedAnswer,
                    Options = q.Options.Select(o => new OptionResponseDTO
                    {
                        Id = o.Id,
                        Text = o.Text,
                        IsCorrect = o.IsCorrect
                    }).ToList()
                }).ToList()
            };
        }

        public async Task<List<QuizSummaryDTO>> GetAllQuizzesAsync()
        {
            var quizzes = await context.Quizzes
                .Include(q => q.Questions)
                .ToListAsync();

            return quizzes.Select(q => new QuizSummaryDTO
            {
                Id = q.Id,
                Title = q.Title,
                Description = q.Description,
                Category = q.Category,
                Difficulty = q.Difficulty,
                LimitTime = q.LimitTime,
                QuestionCount = q.Questions.Count
            }).ToList();
        }

        public async Task<List<QuizSummaryDTO>> GetQuizzesByCategoryAsync(string category)
        {
            var quizzes = await context.Quizzes
                .Include(q => q.Questions)
                .Where(q => q.Category.ToLower() == category.ToLower())
                .ToListAsync();

            return quizzes.Select(q => new QuizSummaryDTO
            {
                Id = q.Id,
                Title = q.Title,
                Description = q.Description,
                Category = q.Category,
                Difficulty = q.Difficulty,
                LimitTime = q.LimitTime,
                QuestionCount = q.Questions.Count
            }).ToList();
        }

        public async Task<List<QuizSummaryDTO>> GetQuizzesByDifficultyAsync(string difficulty)
        {
            var quizzes = await context.Quizzes
                .Include(q => q.Questions)
                .Where(q => q.Difficulty.ToLower() == difficulty.ToLower())
                .ToListAsync();

            return quizzes.Select(q => new QuizSummaryDTO
            {
                Id = q.Id,
                Title = q.Title,
                Description = q.Description,
                Category = q.Category,
                Difficulty = q.Difficulty,
                LimitTime = q.LimitTime,
                QuestionCount = q.Questions.Count
            }).ToList();
        }

        public async Task<QuizResponseDTO?> UpdateQuizAsync(int id, QuizUpdateDTO dto)
        {
            var quiz = await context.Quizzes.FindAsync(id);

            if (quiz == null)
                return null;

            quiz.Title = dto.Title;
            quiz.Description = dto.Description;
            quiz.Category = dto.Category;
            quiz.Difficulty = dto.Difficulty;
            quiz.LimitTime = dto.LimitTime;

            await context.SaveChangesAsync();

            return await GetQuizByIdAsync(id);
        }

        public async Task<bool> DeleteQuizAsync(int id)
        {
            var quiz = await context.Quizzes
                .Include(q => q.Questions)
                    .ThenInclude(q => q.Options)
                .FirstOrDefaultAsync(q => q.Id == id);

            if (quiz == null)
                return false;

            context.Quizzes.Remove(quiz);
            await context.SaveChangesAsync();

            return true;
        }
    }
}

File: frontend/src/services/QuizService.ts
import api from '../api/api.ts';

export interface QuizCreateRequest {
    Title: string;
    Description?: string;
    Category: string;
    Difficulty: string;
    LimitTime: number;
    Questions: QuestionCreateRequest[];
}

export interface QuestionCreateRequest {
    Text: string;
    QuestionType: string;
    Points: number;
    ExpectedAnswer?: string;
    Options: OptionCreateRequest[];
}

export interface OptionCreateRequest {
    Text: string;
    IsCorrect: boolean;
}

export interface QuizResponse {
    Id: number;
    Title: string;
    Description?: string;
    Category: string;
    Difficulty: string;
    LimitTime: number;
    Questions: QuestionResponse[];
}

export interface QuestionResponse {
    Id: number;
    Text: string;
    QuestionType: string;
    Points: number;
    ExpectedAnswer?: string;
    Options: OptionResponse[];
}

export interface OptionResponse {
    Id: number;
    Text: string;
    IsCorrect: boolean;
}

export interface QuizSummary {
    Id: number;
    Title: string;
    Description?: string;
    Category: string;
    Difficulty: string;
    LimitTime: number;
    QuestionCount: number;
}

export const createQuiz = async (data: QuizCreateRequest): Promise<QuizResponse> => {
    const response = await api.post('/api/quiz', data);
    return response.data;
};

export const getAllQuizzes = async (): Promise<QuizSummary[]> => {
    const response = await api.get('/api/quiz');
    return response.data;
};

export const getQuizById = async (id: number): Promise<QuizResponse> => {
    const response = await api.get(`/api/quiz/${id}`);
    return response.data;
};

export const updateQuiz = async (id: number, data: Partial<QuizCreateRequest>): Promise<QuizResponse> => {
    const response = await api.put(`/api/quiz/${id}`, data);
    return response.data;
};

export const deleteQuiz = async (id: number): Promise<void> => {
    await api.delete(`/api/quiz/${id}`);
};

export const getQuizzesByCategory = async (category: string): Promise<QuizSummary[]> => {
    const response = await api.get(`/api/quiz/category/${category}`);
    return response.data;
};

export const getQuizzesByDifficulty = async (difficulty: string): Promise<QuizSummary[]> => {
    const response = await api.get(`/api/quiz/difficulty/${difficulty}`);
    return response.data;
};
